ext/TensorKitChainRulesCoreExt/utility.jl

```diff
@@ -29,3 +29,13 @@ function (::ProjectTo{T1})(x::T2) where {S,N1,N2,T1<:AbstractTensorMap{<:Any,S,N
     end
     return y
 end
+
+function (::ProjectTo{T1})(x::T2) where {T1<:DiagonalTensorMap,T2<:AbstractTensorMap}
+    T1 === T2 && return x
+    y = DiagonalTensorMap{scalartype(T1),spacetype(T1),storagetype(T1)}(undef, space(x, 1))
```

Comment:
Can we make this into
```julia
function (::ProjectTo{DiagonalTensorMap{T,S,A}})(x::AbstractTensorMap) where {T,S,A}
    x isa DiagonalTensorMap{T,S,A} && return x
    V = space(x, 1)
    space(x) == (V ← V) || throw(SpaceMismatch())
    y = DiagonalTensorMap{T,S,A}(undef, V)
    for (c, b) in blocks(y)
         p = ProjectTo(b)
         b .= p(block(x, c))
     end
     return y
 end
```
I think that reads a bit more pleasantly and has a check on the space of `x` being correct.